### PR TITLE
Arm64/Sve: Add FFR register liveness tracking

### PR DESCRIPTION
**Pull request by kunalspathak**
Add support to track liveness for FFR register. 

- The FFR register is defined/written for `SetFFR()` and `LoadVectorFirstFaulting()` APIs
- The FFR register is consumed/read by `GetFFR()` and `LoadVectorFirstFaulting()` APIs.
 
Introduced a special lclNum `lvaFfrRegister` that gets created during lowering whenever we see the first definition of FFR. This `lvaFfrRegister` is used going forward to track various def/use that happen as part of APIs.
I am using `p14` as a `PhysReg` to track the definition of FFR register. For use, if there was a definition of FFR earlier, I add a hidden parameter to the consuming APIs to simulate that the FFR register is used in them. If we did not see definition before the first use, we will just skip the addition of the extra parameter, and hence we won't record a use for it as well.

~All the read/write to FFR register, will be done via `p14`. When we need to spill the FFR, we read the contents of FFR and write them in `p14` via `rdffr` instruction. When we need to unspill, we read the contents back from stack into `p14` and write to FFR using `wrffr` instruction.~

This PR is based off https://github.com/dotnet/runtime/pull/104502 but does not include all the changes from it. I have just extracted out the table entries of the `GetFFr*`, `SetFFr()` and `LoadVectorFirstFaulting` and handling to generate the code.
 
Examples: https://gist.github.com/kunalspathak/b03d740762993482c692dcf7c51d5fa1

### CODE REVIEW

**Comment by kunalspathak** [comment]
@dotnet/jit-contrib @dotnet/arm64-contrib PTAL

File: src/coreclr/jit/codegenarm64.cpp [hunk 1]
```diff
@@ -2954,6 +2954,20 @@ void CodeGen::genCodeForStoreLclVar(GenTreeLclVar* lclNode)
...
+            // We are about to store the FFR on stack. So first read the FFR register and
+            // store it in op1Reg. After that, op1Reg will be stored on the stack.
+            assert(varDsc->lvImplicitlyReferenced == 1);
+
+            regNumber op1Reg = data->GetRegNum();
+            assert(op1Reg != REG_NA);
+            GetEmitter()->emitIns_R(INS_sve_rdffr, EA_SCALABLE, op1Reg);
+        }
+#endif
+
```
Comment (jakobbotsch) [inline src/coreclr/jit/codegenarm64.cpp:2970]: I don't really understand why we need to change codegen for `STORE_LCL_VAR` like this. Shouldn't the `GT_PHYSREG` with `REG_FFR` be what emits `sve_rdffr`?

File: src/coreclr/jit/registerarm64.h [hunk 2]
```diff
@@ -115,7 +115,7 @@ REGDEF(P12,  12+PBASE, PMASK(12),  "p12",  "na")
 REGDEF(P13,  13+PBASE, PMASK(13),  "p13",  "na")
 REGDEF(P14,  14+PBASE, PMASK(14),  "p14",  "na")
 REGDEF(P15,  15+PBASE, PMASK(15),  "p15",  "na")
-
+REGALIAS(FFR, P14)
```
Comment (tannergooding) [inline src/coreclr/jit/registerarm64.h:118]: Why use P14 rather than allocate a new P16 (which closely mirrors how thread suspend tracks it)?

Won't this be problematic since P14 is available for general purpose use otherwise?

Comment (kunalspathak) [reply src/coreclr/jit/registerarm64.h:118]: I started with that approach, but soon realized that it gets assigned to related intervals as well causing us to assign a non-existent register. If I do not make it as part of `RBM_ALLMASK`, we still hit various asserts when we try to create RefPosition or assign register that `P16` is non-existent register and is not part of `RBM_ALLMASK`.

File: src/coreclr/jit/codegenarm64.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:2970]: yes, perhaps i can do that. let me see.

**Review by TIHan — COMMENTED** [review]
The API additions look fine on my end.

File: src/coreclr/jit/registerarm64.h (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/registerarm64.h:118]: It's not clear why this would be an issue here, but not for `REG_EBP`/`REG_ESP`/`REG_K0` (x64) or `REG_IP0`/`REG_IP1`/`REG_PR`/`REG_FP`/`REG_ZR` (arm64). These are usable, but come with special semantics.

Is the issue perhaps that FFR needs to be in `ALLMASK` but not part of the candidate set for normal mask nodes?

Comment (kunalspathak) [reply src/coreclr/jit/registerarm64.h:118]: > Is the issue perhaps that FFR needs to be in ALLMASK but not part of the candidate set for normal mask nodes?

exactly

File: src/coreclr/jit/codegenarm64.cpp (hunk 1 above)
Comment (kunalspathak) [reply src/coreclr/jit/codegenarm64.cpp:2970]: fixed.

File: src/coreclr/jit/registerarm64.h (hunk 2 above)
Comment (tannergooding) [reply src/coreclr/jit/registerarm64.h:118]: Don’t we need to do effectively that here anyways, otherwise LSRA could assign or need pg14 for something else since it is part of the candidate list?

At which point we can just do this without stealing an existing register def away by defining a new pg17?

Comment (kunalspathak) [reply src/coreclr/jit/registerarm64.h:118]: I have already tried that approach. If i have `REG_P16`,  we will allocate that as a register to related intervals too. However, we won't be able to use it directly anyway for storing FFR on stack or for reading the contents out of FFR. To do that, we need additional register and reusing `REG_P14` fits nicely with that mode. `REG_P14` will be busy so it won't get assigned to something else.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 3]
```diff
@@ -2366,6 +2400,43 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
                 break;
             }
 
+            case NI_Sve_LoadVectorFirstFaulting:
+            {
+                if (unspilledFfr)
+                {
+                    // We have unspilled the FFR in op1Reg. Restore it back in FFR register.
+                    GetEmitter()->emitIns_R(INS_sve_wrffr, emitSize, op1Reg, opt);
```
Comment (kunalspathak) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2408]: This should be `op3Reg`. I have fixed this locally.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 4]
```diff
@@ -318,6 +319,39 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+                break;
+            }
+            case NI_Sve_LoadVectorFirstFaulting:
+            {
+                if ((intrin.op3 != nullptr) && ((intrin.op3->gtFlags & GTF_SPILLED) != 0))
+                {
+                    // If there was a op3 for this intrinsic, it means FFR is consumed here
+                    // and we need to unspill.
+                    unspilledFfr = true;
+                }
```
Comment (jakobbotsch) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:349]: So we don't call `genConsumeReg` on this operand? Is `GTF_SPILLED` the only thing that needs to be handled then?

I don't really understand what the FFR register is, but what happens if it gets changed by a callee and we didn't end up spilling the local (that represents p14) around that call. Is that possible?

Comment (jakobbotsch) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:349]: I guess we do call it, and this is detecting if it unspilled into `p14`? I wonder if unspilling itself could have the code to mirror p14 to ffr. I'm sort of worried about what can happen with GT_RELOAD or GT_COPY here... You may want to run jitstressregs (not sure if we'll have any coverage, though?)

Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:349]: >but what happens if it gets changed by a callee and we didn't end up spilling the local (that represents p14) around that call. Is that possible?

That's the entire purpose of this PR, to make sure that we spill it around the call. Since all predicate registers are callee trash, using `P14` does force to save across the call (if it is live).

yes, unspilling will happen as part of `genConsumeMultiOpOperands()`. 

> I wonder if unspilling itself could have the code to mirror p14 to ffr

I could, but that will exercise in the common code path. Having it here already restricts us to do this for scalable APIs which are SVE only.

> I'm sort of worried about what can happen with GT_RELOAD or GT_COPY here... You may want to run jitstressregs (not sure if we'll have any coverage, though?)

Unfortunately we have basic coverage of the CI pipeline that I added few days back that tests with `AltJit` , but not with jitstressregs. I will see if I can test it locally.

Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:349]: i didn't get a good repro, but I can potentially add a `gtSkipReloadOrCopy` here to bypass the COPY/RELOAD to check if i will be unspilling here or not. The `consumeRegs ` will anyway do the right thing.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp [hunk 5]
```diff
@@ -2366,6 +2400,43 @@ void CodeGen::genHWIntrinsic(GenTreeHWIntrinsic* node)
...
+            case NI_Sve_GetFfrInt64:
+            case NI_Sve_GetFfrSByte:
+            case NI_Sve_GetFfrUInt16:
+            case NI_Sve_GetFfrUInt32:
+            case NI_Sve_GetFfrUInt64:
+            {
+                if (unspilledFfr)
+                {
+                    // We have unspilled the FFR in op1Reg. Restore it back in FFR register.
+                    GetEmitter()->emitIns_R(INS_sve_wrffr, emitSize, op1Reg, opt);
```
Comment (jakobbotsch) [inline src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2428]: In this case it seems we could just emit `mov targetReg, op1Reg` and skip `rdffr` entirely?

File: src/coreclr/jit/hwintrinsiclistarm64sve.h [hunk 6]
```diff
@@ -237,6 +246,7 @@ HARDWARE_INTRINSIC(Sve,           Scatter32BitNarrowing,
...
 HARDWARE_INTRINSIC(Sve,           Scatter8BitWithByteOffsetsNarrowing,                              -1,      -1,     false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_sve_st1b,       INS_invalid,        INS_invalid},     HW_Category_MemoryStore,           HW_Flag_Scalable|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialCodeGen|HW_Flag_ExplicitMaskedOperation|HW_Flag_LowMaskedOperation)
+HARDWARE_INTRINSIC(Sve,           SetFfr,                                                           -1,      1,      true,  {INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_sve_wrffr,      INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_ExplicitMaskedOperation|HW_Flag_BaseTypeFromFirstArg|HW_Flag_SpecialSideEffectMask|HW_Flag_SpecialCodeGen)
```
Comment (jakobbotsch) [inline src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: Are you sure `wrffr` is the right instruction to use? I think we should rather be exposing the `setffr` instruction: https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/SETFFR--Initialise-the-first-fault-register-to-all-true-?lang=en

I do not think the JIT should need to be emitting `wrffr` at all. From what I can tell, `wrffr` would be necessary mainly in the kernel around context switching. The manual says: https://developer.arm.com/documentation/ddi0596/2020-12/SVE-Instructions/WRFFR--Write-the-first-fault-register-?lang=en

> Read the source predicate register and place in the first-fault register (FFR). This instruction is intended to restore a saved FFR and is not recommended for general use by applications.
> 
> This instruction requires that the source predicate contains a MONOTONIC predicate value, in which starting from bit 0 there are zero or more 1 bits, followed only by 0 bits in any remaining bit positions. If the source is not a monotonic predicate value, then the resulting value in the FFR will be UNPREDICTABLE. It is not possible to generate a non-monotonic value in FFR when using SETFFR followed by first-fault or non-fault loads.

 For the JIT I think we should do this:
1. Allow the user to emit `setffr` to get `ffr` into a known state before the first-faulting loads
2. Always emit `readffr` after `LoadVectorFirstFaulting` as you are already doing, moving its value into the created local. DCE should be able to remove it if that local does not get read.
3. Emit `readffr` into the local we have after `setffr`, in case the user is reading the "all true" value for some reason
4. Read directly from the local as part of the `GetFfr` functions, emitting no `rdffr` instructions as part of that intrinsic

I think this will allow getting rid of all the special handling around spills.

The only question I have is what the expected behavior is when the user writes code like:
```csharp
SetFfr();
NestedCallThatMayEmitFirstFaultLoads();
GetFfr();
```

I think with the approach in this PR (and the approach I suggest here), the `ffr` will reflect the "all-true" value that `setffr` sets up. But some users may expect the `ffr` to reflect what happened inside the nested call, since `ffr` is a shared resource in the ISA.

We should perhaps document somewhere that the state of `ffr` is not guaranteed to reflect what happened inside nested calls.

Comment (jakobbotsch) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: BTW, I also think it would be just reasonable to not try to save/restore `FFR` around calls in the way we are doing now. These intrinsics really allow the user directly to interact with a resource that is shared in the CPU, so I do not see why we shouldn't just reflect that. That is, we would have `SetFfr()` intrinsic that emits the `setffr` instruction, plus the `GetFfr` intrinsics to emit the `rdffr` instruction. It would be up to the user to make sure they read it with `GetFfr*` intrinsics after doing their first-faulting loads, before the shared resource may be overwritten by e.g. child calls.

File: src/coreclr/jit/hwintrinsiccodegenarm64.cpp (hunk 5 above)
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiccodegenarm64.cpp:2428]: We could, but I will have to double check if that is fine to do, because sometimes, we do want to make sure `ffr` is up-to-date. I will leave this as a future optimization.

File: src/coreclr/jit/hwintrinsiclistarm64sve.h (hunk 6 above)
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: > Why not expose setffr instead of wrffr

I asked this https://github.com/dotnet/runtime/pull/104502#discussion_r1677296041 and we need to revisit `SetFfr()` in API review to see if we can just not pass `mask` parameter, unless @tannergooding knows a use case for having a parameter to `SetFfr()`.

> It would be up to the user to make sure they read it with GetFfr* intrinsics after doing their first-faulting loads, before the shared resource may be overwritten by e.g. child calls. 

This was discussed extensively offline before we came to the consensus of doing it that way. But given the fact that we are exposing `GetFfr()` to the user, we want to make sure the state of FFR is preserved after `LoadFF()` because they will use the FFR contents to see which lanes were read successfully.

Comment (jakobbotsch) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: Thanks, I read the internal email thread and that convinced me that the approach here (to model `FFR` more explicitly per method) is the right one.

Comment (tannergooding) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: It is very explicit that `wrffr` is available for use in user code and that it mutates the register fully. Particularly in C/C++, this is the only way for a developer to utilize FFR and make a call while preserving the state (they are expected to do the saving/restore across the call boundary themselves).

In .NET, we're making it implicit caller save instead so that the JIT can still do its job without incurring unnecessary expense. But that still leaves the need that developers be able to pass back arbitrary FFR state from a helper method and allow the caller to then set FFR to that returned value.

Comment (jakobbotsch) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: It seems like we should expose `wrffr` and `setffr` separately then? The manual clearly directs users to use `setffr`, so it's odd to me that we don't even expose it (in the no-argument version).

Comment (tannergooding) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:249]: We opted to not expose it because the JIT can recognize `WriteFfr(AllBitsSet)` and optimize it to emit `setffr`.

Due to the sheer number of intrinsics that exist we've tried to balance things a little between core functionality and what can be reasonably recognized as a simple IR pattern. We don't expose overloads for all the masked variants of instructions for the same reason, instead recognizing `ConditionalSelect(mask, M(...), mergeValue)`, which has let us cut down on the surface area by several thousand APIs and centralize the support via fairly minimal logic in the JIT instead.

`setffr` is something we still could directly expose, but it's also trivial for the JIT to say `if (op1->IsVectorAllBitsSet()) { MakeSrcContained(node, op1); }` and then `if (op1->IsVectorAllBitsSet()) { ins = INS_setffr; }`

File: src/coreclr/jit/lowerarmarch.cpp [hunk 7]
```diff
@@ -1518,6 +1518,79 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
...
+            break;
+        }
+        case NI_Sve_LoadVectorFirstFaulting:
+        {
+            unsigned lclNum = comp->lvaFfrRegister;
+
+            LIR::Use use;
+            bool     foundUse = BlockRange().TryGetUse(node, &use);
+
+            if (foundUse)
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerarmarch.cpp:1563]: I think we have to create FFR uses/defs even if `!foundUse`. The use is just for the value, but the user could potentially use `LoadVectorFirstFaulting` as a way to probe inaccessible memory as well, without storing the value anywhere.

File: src/coreclr/jit/lowerarmarch.cpp [hunk 8]
```diff
@@ -1518,6 +1518,79 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
...
+        case NI_Sve_GetFfrUInt16:
+        case NI_Sve_GetFfrUInt32:
+        case NI_Sve_GetFfrUInt64:
+        {
+            LIR::Use use;
+            bool     foundUse = BlockRange().TryGetUse(node, &use);
+            if (foundUse)
+            {
+                // Remove GetFfr and replace the GetFfr's use with the lvaFfrRegister lclVar.
+                CreateFFRDefIfNeeded(node);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerarmarch.cpp:1540]: I wouldn't expect this to be necessary here

File: src/coreclr/jit/lowerarmarch.cpp [hunk 9]
```diff
@@ -1518,6 +1518,79 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
...
+
+            if (foundUse)
+            {
+                if (m_ffrTrashed)
+                {
+                    // Consume the FFR register value from local variable to simulate "use" of FFR,
+                    // only if it was trashed. If it was not trashed, we do not have to reload the
+                    // contents of the FFR register.
+
+                    CreateFFRDefIfNeeded(node);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerarmarch.cpp:1571]: I think this has to use `StoreFFRValue`, creating the local if it does not exist. As we discussed offline the block order may not necessarily reveal that there is a `SetFFR` in the function that executed before this node.

File: src/coreclr/jit/hwintrinsiclistarm64sve.h [hunk 10]
```diff
@@ -122,6 +122,14 @@ HARDWARE_INTRINSIC(Sve,           GatherVectorUInt32WithByteOffsetsZeroExtend,
...
+HARDWARE_INTRINSIC(Sve,           GetFfrUInt32,                                                     -1,     -1,      false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_rdffr,      INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_ReturnsPerElementMask|HW_Flag_SpecialSideEffectMask)
+HARDWARE_INTRINSIC(Sve,           GetFfrUInt64,                                                     -1,     -1,      false, {INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_invalid,        INS_sve_rdffr,      INS_invalid,        INS_invalid},     HW_Category_SIMD,                  HW_Flag_Scalable|HW_Flag_SpecialCodeGen|HW_Flag_ReturnsPerElementMask|HW_Flag_SpecialSideEffectMask)
```
Comment (jakobbotsch) [inline src/coreclr/jit/hwintrinsiclistarm64sve.h:132]: I don't think `GetFfr` need `HW_Flag_SpecialSideEffectMask` given how we are modelling this now

File: src/coreclr/jit/registerarm64.h (hunk 2 above)
Comment (jakobbotsch) [reply src/coreclr/jit/registerarm64.h:118]: Registers like the ones you mention are not modelled for spilling at all within LSRA. I do not think it would be that simple to try to model `FFR` as a new register; we end up having to solve what it means to spill and unspill this register, which the JIT only knows how to do to and from stack today. As Kunal mentions that ends up requiring a predicate register anyway, and special spilling logic (much like upper vector saves/restores).

I think the approach taken by this PR (just model the stores/defs to/from a normal mask typed local) is just fine. It means we are essentially manually inserting spills/reloads of the wrffr/rdffr in the IR before LSRA, which means we don't get the benefit of the interference analysis that LSRA has. That part is solvable with a simple analysis that we can do for functions that use FFR. I think that optimization can be done as future work (I don't mind writing that analysis eventually -- ping me when we start implementing BCL methods using these intrinsics...).

File: src/coreclr/jit/hwintrinsiclistarm64sve.h (hunk 10 above)
Comment (kunalspathak) [reply src/coreclr/jit/hwintrinsiclistarm64sve.h:132]: Yes, I need to double check on this flag. Don't remember we had this in first place in Will's PR.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 7 above)
Comment (kunalspathak) [reply src/coreclr/jit/lowerarmarch.cpp:1563]: > as a way to probe inaccessible memory as well

That's a fair point.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 9 above)
Comment (kunalspathak) [reply src/coreclr/jit/lowerarmarch.cpp:1571]: > creating the local if it does not exist

Yes, I need to not have the check for `lclNum == BAD_VAR_NUM` inside `CreateFFRDefIfNeeded()`. `StoreFFRValue()` stores the value "after" the node, while `CreateFFRDefIfNeeded()` needs to store the value "before" the node, so will keep them separate.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 8 above)
Comment (kunalspathak) [reply src/coreclr/jit/lowerarmarch.cpp:1540]: can you elaborate?

Comment (jakobbotsch) [reply src/coreclr/jit/lowerarmarch.cpp:1540]: We insert the stores into `lvaFfrRegister` after all operations that may modify it -- `setffr` and `LoadFF` -- thus the local itself always has the up-to-date value. So I don't think we should read from it as part of `GetFFR` -- it can just use the already up-to-date FFR value directly.

File: src/coreclr/jit/registerarm64.h (hunk 2 above)
Comment (kunalspathak) [reply src/coreclr/jit/registerarm64.h]: yes, I was adding the entry of `REG_FFR` in a wrong spot. Fixed it.

File: src/coreclr/jit/lowerarmarch.cpp (hunk 8 above)
Comment (kunalspathak) [reply src/coreclr/jit/lowerarmarch.cpp]: fixed

File: src/coreclr/jit/instr.cpp [hunk 11]
```diff
@@ -721,7 +721,7 @@ void CodeGen::inst_TT_RV(instruction ins, emitAttr size, GenTree* tree, regNumbe
     unsigned varNum = tree->AsLclVarCommon()->GetLclNum();
     assert(varNum < compiler->lvaCount);
 #if CPU_LOAD_STORE_ARCH
-    assert(GetEmitter()->emitInsIsStore(ins));
+    assert(GetEmitter()->emitInsIsStore(ins) || (ins == INS_sve_str));
```
Comment (jakobbotsch) [inline src/coreclr/jit/instr.cpp:724]: Is this change related?

File: src/coreclr/jit/lowerarmarch.cpp [hunk 12]
```diff
@@ -1518,6 +1518,75 @@ GenTree* Lowering::LowerHWIntrinsic(GenTreeHWIntrinsic* node)
...
+            LIR::Use use;
+            bool     foundUse = BlockRange().TryGetUse(node, &use);
+
+            if (m_ffrTrashed)
+            {
+                // Consume the FFR register value from local variable to simulate "use" of FFR,
+                // only if it was trashed. If it was not trashed, we do not have to reload the
+                // contents of the FFR register.
+
+                CreateFFRDefIfNeeded(node);
```
Comment (jakobbotsch) [inline src/coreclr/jit/lowerarmarch.cpp:1565]: I also think we shouldn't have this one

Comment (kunalspathak) [reply src/coreclr/jit/lowerarmarch.cpp]: ah, yes.

File: src/coreclr/jit/instr.cpp (hunk 11 above)
Comment (kunalspathak) [reply src/coreclr/jit/instr.cpp:724]: It is somewhat related to https://github.com/dotnet/runtime/issues/105512, but I wanted to point fix it here and we can revert this when we properly fix https://github.com/dotnet/runtime/issues/105512.

File: src/coreclr/jit/simd.cpp [hunk 13]
```diff
@@ -130,6 +130,29 @@ unsigned Compiler::getSIMDInitTempVarNum(var_types simdType)
...
+//    simdType - Required SIMD type
+//
+// Returns:
+//    The temp number
+//
+unsigned Compiler::getFFRegisterVarNum()
+{
+    if (lvaFfrRegister == BAD_VAR_NUM)
+    {
+        lvaFfrRegister                  = lvaGrabTempWithImplicitUse(false DEBUGARG("Save the FFR value."));
```
Comment (jakobbotsch) [inline src/coreclr/jit/simd.cpp:148]: I'm not sure if it matters, but this does not need to be "implicit use" anymore I think.

**Review by jakobbotsch — APPROVED** [review]
LGTM

File: src/coreclr/jit/simd.cpp (hunk 13 above)
Comment (kunalspathak) [reply src/coreclr/jit/simd.cpp]: yes, I had this when I was implementing this in importer. don't need it now.
